src/util/shm/mpir_shm_coll.c

```diff
@@ -3,36 +3,156 @@
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIR_SHM_COLL_FINALIZE);
     return mpi_errno;
+  fn_fail:
+    goto fn_exit;
 }
 
 int MPIR_SHM_bcast(void *buffer, int count, MPI_Datatype datatype, int root)
 {
     int mpi_errno = MPI_SUCCESS;
+    MPI_Aint size, len;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_SHM_BCAST);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_SHM_BCAST);
+
+    /* We only support builtin datatypes */
+    MPIR_Assert(HANDLE_GET_KIND(datatype) == HANDLE_KIND_BUILTIN);
+
+    MPIR_Datatype_get_size_macro(datatype, size);
+    len = size * count;
+    MPIR_Assert(len <= PROC_MEM_LEN);
+
+    if (local_rank == root) {
+        MPIR_Memcpy(baseaddr + root * PROC_MEM_LEN, buffer, len);
+        mpi_errno = MPIR_shm_barrier(barrier, local_size);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+    } else {
+        mpi_errno = MPIR_shm_barrier(barrier, local_size);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+        MPIR_Memcpy(buffer, baseaddr + root * PROC_MEM_LEN, len);
+    }
+
+    mpi_errno = MPIR_shm_barrier(barrier, local_size);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIR_SHM_BCAST);
     return mpi_errno;
+  fn_fail:
+    goto fn_exit;
 }
 
 int MPIR_SHM_allgather(const void *sendbuf, int sendcount, MPI_Datatype sendtype, void *recvbuf,
                        int recvcount, MPI_Datatype recvtype)
 {
     int mpi_errno = MPI_SUCCESS;
+    MPI_Aint size, len;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_SHM_ALLGATHER);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_SHM_ALLGATHER);
+
+    /* We only support builtin datatypes */
+    MPIR_Assert(HANDLE_GET_KIND(sendtype) == HANDLE_KIND_BUILTIN);
+    MPIR_Assert(HANDLE_GET_KIND(recvtype) == HANDLE_KIND_BUILTIN);
+
+    /* We only allow send and recv to be identical */
+    MPIR_Datatype_get_size_macro(sendtype, size);
+    len = size * sendcount;
+    MPIR_Assert(len <= PROC_MEM_LEN);
+    MPIR_Assert(sendtype == recvtype);
+    MPIR_Assert(sendcount == recvcount);
+
+    const void *buf =
+        (sendbuf == MPI_IN_PLACE) ? (const void *) (recvbuf + local_rank * len) : sendbuf;
+
+    /* Every proc copies its data in shm region */
+    MPIR_Memcpy(baseaddr + local_rank * PROC_MEM_LEN, buf, len);
+    mpi_errno = MPIR_shm_barrier(barrier, local_size);
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    /* Gather data from all procs */
+    for (int i = 0; i < local_size; i++)
+        MPIR_Memcpy(recvbuf + i * len, baseaddr + i * PROC_MEM_LEN, len);
+
+    mpi_errno = MPIR_shm_barrier(barrier, local_size);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIR_SHM_ALLGATHER);
     return mpi_errno;
+  fn_fail:
+    goto fn_exit;
 }
 
 int MPIR_SHM_barrier(void)
 {
     int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_SHM_BARRIER);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_SHM_COLL_BARRIER);
+
+    mpi_errno = MPIR_shm_barrier(barrier, local_size);
```

Comment:
These names confused me a little.. `MPIR_SHM_barrier` and `MPIR_shm_barrier`

Reply:
@jain-surabhi-23 the idea here is that with these collectives we don't need to use `MPIDU_shm_seg_(alloc,commit)` anymore. These also allocate a shared memory barrier for every new segment which is wasteful. Additionally, they also return other unneeded information that could just be hidden in the alloc function. `MPIDU_shm_seg_(alloc,commit)` will be replaced by a `MPIR_SHM_alloc` which will use the collectives internally for the descriptor exchange among processes. This new alloc function should only return a pointer to the allocated shared memory, no pointer to any barrier. The `MPIDU_shm_barrier` implementation will become static inside `mpir_shm_coll.c` and other layers will just use it.

Reply:
@jain-surabhi-23 I think @hzhou also had some comments about the actual necessity of using collective semantics at all and proposed to just expose a shared memory segment, initialized first, into which processes can place their data later on. As the current implementation of collectives does not do anything too fancy this proposal makes sense to me. However, if we want to support larger buffers without allocating a lot of memory we can add the additional logic inside the collectives. I would like to hear some other opinions as well @raffenet @pavanbalaji

Reply:
I think just making a segment available at init time makes sense. I think the allgather would be a convenience, but if the buffer is statically allocated, it shouldn't be too hard to know where to place data. Just to be clear, we would still retain the barrier operation, right?

Reply:
@raffenet yes that would be allocated once with the initial segment and made available to other components to be used through `MPIR_SHM_barrier()`

Reply:
Got it. I think in the case of things like business card exchange, shm collectives, and win allocate shared, there is still need to allocate additional shm segments. This init-time segment would be useful for those creations by allowing us to bypass PMI and synchronize processes at the node-level.